### ISSUE TITLE
Nagios Grok pattern matching seems to be broken in Ingest node processor

### ISSUE THREAD

**Issue by jlecour** [elastic/elasticsearch#22782]
**Elasticsearch version**: 5.1.2

**Plugins installed**: []

**JVM version**: 1.8.0_102

**OS version**: Debian 8.7.1

**Description of the problem including expected versus actual behavior**:

I'm trying to move my logstash basic pipelines to a more direct Ingest node setup.

The `NAGIOSLOGLINE` is a core pattern provided by Elasticsearch in the Ingest node family of processors.

In the first example, I've overwritten the pattern to simplify it at the extreme and show the simplest expected behavior.

It is using regular existing sub-patterns (they are here : https://raw.githubusercontent.com/elastic/elasticsearch/master/modules/ingest-common/src/main/resources/patterns/nagios)

Request : 

```js
{
  "pipeline": {
    "processors" : [
      {
        "grok": {
          "field": "message",
          "patterns": ["%{NAGIOSLOGLINE}"],
          "pattern_definitions" : {
            "NAGIOSLOGLINE": "%{NAGIOSTIME} (?:%{NAGIOS_CURRENT_SERVICE_STATE})"
          }
        }
      },
      {
        "remove": {
          "field": "message"
        }
      }
    ]
  },
  "docs": [
    {
      "_source": {
        "message": "[1466460000] CURRENT SERVICE STATE: myhostname;myservice;CRITICAL;HARD;3;DISK CRITICAL - free space: / 554 MB (62% inode=78%): /run 100 MB (99% inode=99%): /run/lock 5 MB (100% inode=99%): /run/shm 200 MB (100% inode=99%): /boot 395 MB (90% inode=99%): /home 2992 MB (2% inode=94%): /tmp 426 MB (97% inode=99%): /usr 3435 MB (77% inode=84%): /var 4180 MB (46% inode=97%):"
      }
    }
  ]
}

```

Response (perfectly matched) : 

```js
    {
        "docs":
        [
            {
                "doc":
                {
                    "_type": "_type",
                    "_id": "_id",
                    "_index": "_index",
                    "_source":
                    {
                        "nagios_message": "DISK CRITICAL - free space: / 554 MB (62% inode=78%): /run 100 MB (99% inode=99%): /run/lock 5 MB (100% inode=99%): /run/shm 200 MB (100% inode=99%): /boot 395 MB (90% inode=99%): /home 2992 MB (2% inode=94%): /tmp 426 MB (97% inode=99%): /usr 3435 MB (77% inode=84%): /var 4180 MB (46% inode=97%):",
                        "nagios_epoch": "1466460000",
                        "nagios_statecode": "3",
                        "nagios_hostname": "myhostname",
                        "nagios_state": "CRITICAL",
                        "nagios_statetype": "HARD",
                        "nagios_service": "myservice",
                        "nagios_type": "CURRENT SERVICE STATE"
                    },
                    "_ingest":
                    {
                        "timestamp": "2017-01-25T10:00:01.671+0000"
                    }
                }
            }
        ]
    }
```

Here is the request, using the top level pattern : 

```js
{
  "pipeline": {
    "processors" : [
      {
        "grok": {
          "field": "message",
          "patterns": ["%{NAGIOSLOGLINE}"]
        }
      },
      {
        "remove": {
          "field": "message"
        }
      }
    ]
  },
  "docs": [
    {
      "_source": {
        "message": "[1466460000] CURRENT SERVICE STATE: myhostname;myservice;CRITICAL;HARD;3;DISK CRITICAL - free space: / 554 MB (62% inode=78%): /run 100 MB (99% inode=99%): /run/lock 5 MB (100% inode=99%): /run/shm 200 MB (100% inode=99%): /boot 395 MB (90% inode=99%): /home 2992 MB (2% inode=94%): /tmp 426 MB (97% inode=99%): /usr 3435 MB (77% inode=84%): /var 4180 MB (46% inode=97%):"
      }
    }
  ]
}
```

Here is the response, where many fields are missing : 

```js
    {
        "docs":
        [
            {
                "doc":
                {
                    "_type": "_type",
                    "_id": "_id",
                    "_index": "_index",
                    "_source":
                    {
                        "nagios_epoch": "1466460000",
                        "nagios_statecode": "3",
                        "nagios_state": "CRITICAL",
                        "nagios_statetype": "HARD"
                    },
                    "_ingest":
                    {
                        "timestamp": "2017-01-25T10:05:21.804+0000"
                    }
                }
            }
        ]
    }
```

With trial and error I've found that the very complex regular expression, with its many sub-patterns, might swallow fields.
It is working fine in Logstash (in production for months) but fails miserably in Elasticsearch. Maybe it's because of a different implementation of the Grok between Logstash and Elasticsearch.

I didn't find any unit test for Nagios grok patterns :/

**Comment by talevy** [comment]
After looking at the nagios grok field matching, I believe those fields are being consumed due to the first pattern in a string of `|`(or) matches fails to match. Might be related to this: https://github.com/elastic/elasticsearch/pull/22131.

I will try this out on the ES 5.2 candidate that will be released soon

**Comment by talevy** [comment]
Just tried it out with 5.2,

I get back this response from the simulate request with the distributed `NAGIOSLOGLINE` grok pattern.

```json
{
  "docs" : [
    {
      "doc" : {
        "_id" : "_id",
        "_type" : "_type",
        "_index" : "_index",
        "_source" : {
          "nagios_message" : "DISK CRITICAL - free space: / 554 MB (62% inode=78%): /run 100 MB (99% inode=99%): /run/lock 5 MB (100% inode=99%): /run/shm 200 MB (100% inode=99%): /boot 395 MB (90% inode=99%): /home 2992 MB (2% inode=94%): /tmp 426 MB (97% inode=99%): /usr 3435 MB (77% inode=84%): /var 4180 MB (46% inode=97%):",
          "nagios_epoch" : "1466460000",
          "nagios_statecode" : "3",
          "nagios_hostname" : "myhostname",
          "nagios_state" : "CRITICAL",
          "nagios_statetype" : "HARD",
          "nagios_service" : "myservice",
          "nagios_type" : "CURRENT SERVICE STATE"
        },
        "_ingest" : {
          "timestamp" : "2017-01-25T18:13:23.036Z"
        }
      }
    }
  ]
}
```

@jlecour, looks like it was fixed in the next release. Mind verifying once Elasticsearch 5.2 is out so we can close the loop on this?

thanks!

**Comment by jlecour** [comment]
Hi and thank you for having looked into this quickly.

Is there an easy way to install the 5.2 candaidate ? If so, I could try on my exact setup and verify that it's fixed.

Thanks

**Comment by talevy** [comment]
@jlecour 

here is a snapshot: https://snapshots.elastic.co/downloads/elasticsearch/elasticsearch-5.2.0-SNAPSHOT.tar.gz

Also, 5.2 should be released really soon!

**Comment by jlecour** [comment]
@talevy 

It's fixed in 5.2 \o/

**Comment by clintongormley** [comment]
thanks for letting us know @jlecour
